pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go

```diff
@@ -204,50 +318,102 @@ func setDefaultDeploymentSpec(spec *appsv1.DeploymentSpec) {
-		return nil, err
+// Function to add a new environment variable to a specific container in the DeploymentSpec
+func addEnvVarToDeploymentSpec(deploymentSpec *appsv1.DeploymentSpec, containerName, envName, envValue string) {
+	// Iterate over the containers in the PodTemplateSpec to find the specified container
+	for i, container := range deploymentSpec.Template.Spec.Containers {
+		if container.Name == containerName {
+			if _, exists := utils.GetEnvVarValue(container.Env, envName); exists {
+				// Overwrite the environment variable
+				for j, envVar := range container.Env {
+					if envVar.Name == envName {
+						deploymentSpec.Template.Spec.Containers[i].Env[j].Value = envValue
+						break
+					}
+				}
+			} else {
+				// Add the new environment variable to the Env field if it ooes not exist
+				container.Env = append(container.Env, corev1.EnvVar{
+					Name:  envName,
+					Value: envValue,
+				})
+				deploymentSpec.Template.Spec.Containers[i].Env = container.Env
+			}
+			log.Info("Added env variable to container",
+				"envName", envName,
+				"envValue", envValue,
+				"containerName", containerName)
+			return
+		}
 	}
-	log.Info("deployment reconcile", "checkResult", checkResult, "err", err)
+	log.Info("Container not found in DeploymentSpec", "containerName", containerName)
+}
 
-	var opErr error
-	switch checkResult {
-	case constants.CheckResultCreate:
-		opErr = r.client.Create(context.TODO(), r.Deployment)
-	case constants.CheckResultUpdate:
-		curJson, err := json.Marshal(deployment)
-		if err != nil {
-			return nil, err
-		}
+func addGPUResourceToDeployment(deployment *appsv1.Deployment, targetContainerName string, tensorParallelSize string) {
+	for i, container := range deployment.Spec.Template.Spec.Containers {
+		if container.Name == targetContainerName {
+			// Initialize Limits map if it's nil
+			if container.Resources.Limits == nil {
+				deployment.Spec.Template.Spec.Containers[i].Resources.Limits = make(map[corev1.ResourceName]resource.Quantity)
+			}
 
-		// To avoid the conflict between HPA and Deployment,
-		// we need to remove the Replicas field from the deployment spec
-		modDeployment := r.Deployment.DeepCopy()
-		modDeployment.Spec.Replicas = nil
+			// Assign the tensorParallelSize value to the GPU "nvidia.com/gpu" resource limits
+			deployment.Spec.Template.Spec.Containers[i].Resources.Limits[constants.NvidiaGPUResourceType] = resource.MustParse(tensorParallelSize)
 
-		modJson, err := json.Marshal(modDeployment)
-		if err != nil {
-			return nil, err
+			// Initialize Requests map if it's nil
+			if container.Resources.Requests == nil {
+				deployment.Spec.Template.Spec.Containers[i].Resources.Requests = make(map[corev1.ResourceName]resource.Quantity)
+			}
+
+			// Assign the tensorParallelSize value to the GPU "nvidia.com/gpu" resource requests
+			deployment.Spec.Template.Spec.Containers[i].Resources.Requests[constants.NvidiaGPUResourceType] = resource.MustParse(tensorParallelSize)
```

Comment:
here you are considering that this method would not be used by anyone else than multi-node feature.
is it safe to assume that?

Reply:
This method is not exclusively for multi-node use, so I'm not sure why you would think that. It sets the GPU value in the resource of a specific container within the deployment containers to the value of tensor-parallel-size

Reply:
Where do you check if the container you are adding the GPU requests is the worker nodes?

Reply:
This is in a method so we should see where calls this method. You can find [the caller ](https://github.com/kserve/kserve/pull/3972/files#diff-401b6cf36f71bdda9ab1593723f98632b0e7ee8d3a98151d65439e13d8d2ea9fR90)

This is one of the callers :
```
addGPUResourceToDeployment(defaultDeployment, constants.InferenceServiceContainerName, tensorParallelSize)
```
the second parameter show which container name need this GPU resource. In this case, it is `kserve-container` so it will be default deployment.

another caller is [here](https://github.com/kserve/kserve/pull/3972/files#diff-401b6cf36f71bdda9ab1593723f98632b0e7ee8d3a98151d65439e13d8d2ea9fR120-R127)
```
if multiNodeEnabled {
		workerDeployment := createRawWorkerDeployment(workerComponentMeta, componentExt, workerPodSpec, componentMeta.Name, pipelineParallelSize, isvcName)

		// Update GPU resource of workerPodSpec based on tensor-parallelSize
		addGPUResourceToDeployment(workerDeployment, constants.WorkerContainerName, tensorParallelSize)
		deploymentList = append(deploymentList, workerDeployment)
	}
```
It is called when multinode is enabled and it also checks if the container name(` constants.WorkerContainerName` is worker node or not.